pkg/compactor/compactor.go

```diff
@@ -0,0 +1,154 @@
+	return &Compactor{
+		logger: logger,
+		store:  store,
+	}
+}
+
+// Compact is mainly responsible for applying snapshots (full + delta), compacting, drefragmenting, taking the snapshot and saving it sequentially.
+func (cp *Compactor) Compact(ro *brtypes.RestoreOptions, needDefragmentation bool) (*brtypes.Snapshot, error) {
+	cp.logger.Info("Start compacting")
+
+	// Deepcopy restoration options ro to avoid any mutation of the passing object
+	cmpctOptions := ro.DeepCopy()
+
+	// If no basesnapshot is found, abort compaction as there would be nothing to compact
+	if cmpctOptions.BaseSnapshot == nil {
+		cp.logger.Error("No base snapshot found. Nothing is available for compaction")
+		return nil, fmt.Errorf("no base snapshot found. Nothing is available for compaction")
+	}
+
+	// Set a temporary etcd data directory for embedded etcd
+	cmpctDir, err := ioutil.TempDir("/tmp", "compactor-")
+	if err != nil {
+		cp.logger.Errorf("Unable to create temporary etcd directory for compaction: %s", err.Error())
+		return nil, err
+	}
+
+	defer os.RemoveAll(cmpctDir)
+
+	cmpctOptions.Config.RestoreDataDir = cmpctDir
+
+	// Then restore from the snapshots
+	r := restorer.NewRestorer(cp.store, cp.logger)
+	embeddedEtcd, err := r.Restore(*cmpctOptions)
+	if err != nil {
+		return nil, fmt.Errorf("unable to restore snapshots during compaction: %v", err)
+	}
+
+	cp.logger.Info("Restoration for compaction is over")
+	// There is a possibility that restore operation may not start an embedded ETCD.
+	if embeddedEtcd == nil {
+		embeddedEtcd, err = miscellaneous.StartEmbeddedEtcd(cp.logger, cmpctOptions)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	defer func() {
+		embeddedEtcd.Server.Stop()
+		embeddedEtcd.Close()
+	}()
+
+	ep := []string{embeddedEtcd.Clients[0].Addr().String()}
+
+	// Then compact ETCD
+
+	// Build Client
+	cfg := clientv3.Config{MaxCallSendMsgSize: cmpctOptions.Config.MaxCallSendMsgSize, Endpoints: ep}
+	client, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build client")
+	}
+	defer client.Close()
+
+	// Build contxt
+	ctx := context.TODO()
+	revCheckCtx, cancel := context.WithTimeout(ctx, etcdDialTimeout)
+	getResponse, err := client.Get(revCheckCtx, "foo")
+	cancel()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to client: %v", err)
+	}
+	etcdRevision := getResponse.Header.GetRevision()
+
+	// Compact
+	if _, err := client.Compact(ctx, etcdRevision); err != nil {
+		return nil, fmt.Errorf("failed to compact: %v", err)
+	}
+
+	// Then defrag the ETCD
+	if needDefragmentation {
+		etcdDialCtx, etcdDialCancel := context.WithTimeout(ctx, etcdutil.EtcdDialTimeout)
+		defragCtx, defragCancel := context.WithTimeout(ctx, etcdutil.DefaultDefragConnectionTimeout)
+		err := etcdutil.DefragmentData(etcdDialCtx, defragCtx, client, ep, cp.logger)
+		etcdDialCancel()
+		defragCancel()
+		if err != nil {
+			cp.logger.Errorf("failed to defragment: %v", err)
+		}
```

Comment:
I meant to pass only `defragCtx` as the context. Do we really need to pass `etcdDialCtx`? Can't we use the same context for both getting the status and triggering defragmentation?